src/libraries/Common/src/System/Reflection/TypeNameParser.cs

```diff
@@ -0,0 +1,632 @@
+
+            if (sb.Length == 0)
+            {
+                // The identifier has to be non-empty
+                _errorIndex = src;
+                ParseError();
+                return null;
+            }
+
+            return sb.ToString();
+        }
+
+        //
+        // Lex the next segment as the assembly name at the end of an assembly-qualified type name. (Do not use for
+        // assembly names embedded inside generic type arguments.)
+        //
+        private string GetNextAssemblyName()
+        {
+            SkipWhiteSpace();
+
+            string assemblyName = new string(_input.Slice(_index));
+            _index = _input.Length;
+            return assemblyName;
+        }
+
+        //
+        // Lex the next segment as an assembly name embedded inside a generic argument type.
+        //
+        // Terminated by an unescaped ']'.
+        //
+        private string? GetNextEmbeddedAssemblyName()
+        {
+            SkipWhiteSpace();
+
+            ValueStringBuilder sb = new ValueStringBuilder(stackalloc char[64]);
+
+            int src = _index;
+            while (true)
+            {
+                if (src >= _input.Length)
+                {
+                    ParseError();
+                    return null;
+                }
+                char c = _input[src];
+                if (c == ']')
+                    break;
+                src++;
+
+                // Backslash can be used to escape a ']' - any other backslash character is left alone (along with the backslash)
+                // for the AssemblyName parser to handle.
+                if (c == '\\' && (src < _input.Length) && _input[src] == ']')
+                {
+                    c = _input[src++];
+                }
+                sb.Append(c);
+            }
+            _index = src;
+
+            if (sb.Length == 0)
+            {
+                // The assembly name has to be non-empty
+                _errorIndex = src;
+                ParseError();
+                return null;
+            }
+
+            return sb.ToString();
+        }
+
+        //
+        // Classify a character as a TokenType. (Fortunately, all tokens in type name strings other than identifiers are single-character tokens.)
+        //
+        private static TokenType CharToToken(char c)
+        {
+            switch (c)
+            {
+                case '\0':
+                    return TokenType.End;
+                case '[':
+                    return TokenType.OpenSqBracket;
+                case ']':
+                    return TokenType.CloseSqBracket;
+                case ',':
+                    return TokenType.Comma;
+                case '+':
+                    return TokenType.Plus;
+                case '*':
+                    return TokenType.Asterisk;
+                case '&':
+                    return TokenType.Ampersand;
+                default:
+                    return TokenType.Other;
+            }
```

Comment:
```suggestion
            return c switch
            {
                '\0' => TokenType.End,
                '[' => TokenType.OpenSqBracket,
                ']' => TokenType.CloseSqBracket,
                ',' => TokenType.Comma,
                '+' => TokenType.Plus,
                '*' => TokenType.Asterisk,
                '&' => TokenType.Ampersand,
                _ => TokenType.Other,
            };
```

Reply:
This is one of those cases where this syntax is less verbose and seems more appropriate.